tests/mssqltestutils.py

```diff
@@ -90,29 +92,71 @@ def getTempPath(*args):
     return  os.path.abspath(tempPath)
 
 def create_test_db():
+    """
+    Creates database for test, using various status checks and retry logic for reliability.
+    - Calls helper method to check status of create db task, if possible
+    - Exits on successful response or retry period exceeds time limit
+    """
     client = create_mssql_cli_client()
     local_machine_name = socket.gethostname().replace(
         '-', '_').replace('.', '_')
+
     test_db_name = u'mssqlcli_testdb_{0}_{1}'.format(
         local_machine_name, random_str())
     query_db_create = u"CREATE DATABASE {0};".format(test_db_name)
-    count = 0
 
-    # retry logic in case db create fails
-    while count < 5:
-        for _, _, status, _, is_create_error in client.execute_query(query_db_create):
-            if not is_create_error:
+    for _, _, status, _, is_create_error in client.execute_query(query_db_create):
+        try:
+            create_db_status = check_create_db_status(test_db_name, client)
+
+            if is_create_error or create_db_status == 'FAILED':
+                # log warning to console and cleanup db
+                warnings.warn('Test DB create failed with error: {0}'.format(status))
+                clean_up_test_db(test_db_name)
+
+            elif create_db_status == 'COMPLETED':
                 shutdown(client)
                 return test_db_name
-            # log warning to console
-            warnings.warn('Test DB create failed with error: {0}'.format(status))
-        count += 1
+        except PermissionError:
+            # PermissionError is returned if the check_create_db_status function creates db
+            # using a non-master db. only master db is supported.
+            shutdown(client)
+            warnings.warn(UserWarning("Warning: use the master DB with your SQL Server to "
+                                      "create test DBs more reliably."))
+            return test_db_name
+
     shutdown(client)
 
     # cleanup db just in case, then raise exception
     clean_up_test_db(test_db_name)
     raise AssertionError("DB creation failed.")
 
+def check_create_db_status(db_name, client):
+    """
+    Uses retry logic with sys.dm_operation_status to check statis of create database job.
+    """
+    if client.database.lower() != 'master':
```

Comment:
This looks good, but can we move this to a higher level? Seems like this check should occur before we issue the 'create database statement'.